src/app/public/src/modules/page-anchor/page-anchor.component.ts

```diff
@@ -1,109 +1,118 @@
+    this.fragment = this.getFragment(this.name, anchorId);
+    this.path = [this.routerService.getActiveUrl()];
 
-    return tutorialAnchorOffsetElement ?
-      tutorialAnchorOffsetElement.offsetTop :
-      this.elementRef.nativeElement.offsetTop;
-  }
+    this.anchor.next({
+      path: this.path,
+      name: this.name,
+      fragment: this.fragment,
+      offsetTop: this.getValidOffsetTop(this.elementRef.nativeElement)
+    } as StacheNavLink);
 
-  private getOrder(): void {
-    let anchors = this.windowRef.nativeWindow.document.querySelectorAll('stache-page-anchor div');
-    for (let i = 0; i < anchors.length; i++) {
-      if (this.fragment === anchors[i].id) {
-        this.order = i;
-      }
-    }
+    const pageAnchors = this.anchorService.pageAnchors.getValue();
+    pageAnchors.push(this.anchor);
+    this.sortPageAnchors(pageAnchors);
+    this.anchorService.pageAnchors.next(pageAnchors);
   }
 
-  private registerAnchor(): void {
-    this.anchorService.addPageAnchor({
+  private updatePageAnchor() {
+    this.anchor.next({
       path: this.path,
       name: this.name,
       fragment: this.fragment,
-      order: this.order,
-      offsetTop: this.offsetTop
+      offsetTop: this.getValidOffsetTop(this.elementRef.nativeElement)
     } as StacheNavLink);
   }
 
-  private findAncestor(element: HTMLElement, elClass: string): HTMLElement {
-    while (element) {
-      if (element.classList.contains(elClass)) {
-        return element;
+  // Sometimes offsetTop will be a sub-zero number and will ruin page tracking sync
+  // This function runs up the parent tree until it finds an element that has a useable offset value
+  private getValidOffsetTop(element: any): number {
+    const tutorialOffsetTop = this.getTutorialOffsetTop(element);
+    if (tutorialOffsetTop) {
+      return tutorialOffsetTop;
+    } else {
+      if (element.offsetTop < 1 && element.parentElement) {
+        return this.getValidOffsetTop(element.parentElement);
       }
-      element = element.parentElement;
+      return element.offsetTop;
     }
-    return undefined;
+  }
+
+  // The Tutorial components have styles that alter how offsetTop is calculated on elements.
+  // This step checks to see if an anchor is sitting in a tutorial block and gets the offsetTop of
+  // that element, which is accurate.
+  private getTutorialOffsetTop(element: any): number {
+    if (!element.classList.contains('stache-tutorial-step') && element.parentElement) {
+      return this.getTutorialOffsetTop(element.parentElement);
+    } else if (element.classList.contains('stache-tutorial-step')) {
+      return element.offsetTop;
+    }
+  }
+
+  private getName(element: any): string {
+    return element.textContent.trim();
+  }
+
+  private getFragment(name: string, anchorId: string): string {
+    return anchorId || name
+      .toLowerCase()
+      .replace(/ /g, '-')
+      .replace(/[^\w-]+/g, '');
+  }
+
+  // TODO: This might not sort properly, must check
```

Comment:
Why are we sorting all the page anchors within the page anchors instead of in the anchor service?  I would suggest either the service, or the subscriber of the `pageAnchors[]` in this case the wrapper, be responsible for sorting?